### PR TITLE
Allow show to accept object

### PR DESCRIPTION
**Pull request by taysea**
#### What does this PR do?
Updates `show` for DataTable, List, and Pagination. Props should always be interpreted from the perspective of the component they are being applied to.

For DataTable and List:
- `show` by default means what **item** should be in view. If a user would like to specify a page to show, they can say `show={{page: 5}}`, for example.

For Pagination:
- `show` by default means what **page** should be in view. If a user would like to specify a child to show, they can say `show={{item: 4}}`, for example.

#### Where should the reviewer start?
src/js/components/Pagination/Pagination.js

#### What testing has been done on this PR?
Tested in storybooks, will make sure we add this to list of stories to be written/tests to be written.

#### How should this be manually tested?
Apply show to any of the above mentioned components.

### CODE REVIEW

File: src/js/components/Pagination/Pagination.js [hunk 1]
```diff
@@ -31,9 +31,23 @@ const Pagination = forwardRef(
     ref,
   ) => {
     const theme = useContext(ThemeContext) || defaultProps.theme;
+
+    let show;
+    let showItem;
+    // if children are provided, show can take form of { item: # },
+    // where # is the item to show
+    if (typeof showProp === 'object' && 'item' in showProp) {
```
Comment (halocline) [inline src/js/components/Pagination/Pagination.js:39]: What if I pass `show={{page: 3}}`? Should we handle that case? `show = showProp.page`

File: src/js/components/DataTable/DataTable.js [hunk 2]
```diff
@@ -193,9 +193,19 @@ const DataTable = ({
     console.warn('DataTable cannot combine "size" and "resizeble".');
   }
 
+  let show;
+  let showItem;
+  // by default, show refers to an item in the DataTable,
+  // but if using pagination, show can take the form of { page: # },
+  // where number refers to the page # to show
+  if (typeof showProp === 'number') showItem = showProp;
+  else if (typeof showProp === 'object' && 'page' in showProp) {
```
Comment (halocline) [inline src/js/components/DataTable/DataTable.js:202]: handle `showProp.item` as well?

File: src/js/components/List/List.js [hunk 3]
```diff
@@ -86,9 +86,19 @@ const List = React.forwardRef(
     const [active, setActive] = useState();
     const [itemFocus, setItemFocus] = useState();
 
+    let show;
+    let showItem;
+    // by default, show refers to an item in the List,
+    // but if using pagination, show can take the form of { page: # },
+    // where number refers to the page # to show
+    if (typeof showProp === 'number') showItem = showProp;
+    else if (typeof showProp === 'object' && 'page' in showProp) {
```
Comment (halocline) [inline src/js/components/List/List.js:95]: handle `showProp.item` as well?

**Review by halocline — COMMENTED** [review]
Wondering if we can support all 3 cases no matter the context?
- show=#
- show={{page: #}}
- show={{item: #}}

What about extending `pagination.js` with a normalize-like function?

const normalizeShow = (show: number | object, context: oneOfType(['pagination', 'dataTable', 'list'])) => {
  // logic based on context
  return normalizedShow
}

This could be done similarly to how DataTable Cell.js adjusts based off of 'header', 'footer', etc. contexts.

File: src/js/components/Pagination/Pagination.js (hunk 1 above)
Comment (taysea) [reply src/js/components/Pagination/Pagination.js:39]: I like the option of handling both. I'll update.

File: src/js/components/List/List.js (hunk 3 above)
Comment (taysea) [reply src/js/components/List/List.js]: Done.

File: src/js/components/DataTable/DataTable.js (hunk 2 above)
Comment (taysea) [reply src/js/components/DataTable/DataTable.js]: Done.

File: src/js/components/Pagination/Pagination.js (hunk 1 above)
Comment (taysea) [reply src/js/components/Pagination/Pagination.js]: Done.

File: src/js/utils/pagination.js [hunk 4]
```diff
@@ -1,13 +1,57 @@
 import { useCallback, useEffect, useState } from 'react';
 
+export const normalizeShow = (showProp, context) => {
+  let show;
+  let showItem;
+
+  if (context === 'dataTable' || context === 'list') {
```
Comment (halocline) [inline src/js/utils/pagination.js:7]:
```suggestion
  if (['dataTable', 'list'].includes(context)) {
```

File: src/js/utils/pagination.js [hunk 5]
```diff
@@ -1,13 +1,57 @@
...
+    // by default, show refers to the index of an item in the List,
+    // but if using pagination, show can take the form of { page: # },
+    // where page refers to the page # to show or { index: # },
+    // where index refers to the index of an item in the list
+    if (typeof showProp === 'number') showItem = showProp;
+    else if (typeof showProp === 'object') {
+      // if user provides both page and index to showProp, index should win
```
Comment (halocline) [inline src/js/utils/pagination.js:14]:
```suggestion
      // in this context, if user provides both page and index to showProp, index should win
```

File: src/js/utils/pagination.js [hunk 6]
```diff
@@ -1,13 +1,57 @@
...
+      } else if ('page' in showProp) show = showProp.page;
+    }
+  } else if (context === 'pagination') {
+    if (typeof showProp === 'number') show = showProp;
+    // if children are provided, show can take form of { index: # },
+    // where index refers to the index of a child to show or { page: # },
+    // where page refers to the page # to show
+    else if (typeof showProp === 'object') {
+      if ('page' in showProp) {
+        // if user provides both page and item to showProp, page should win
```
Comment (halocline) [inline src/js/utils/pagination.js:26]:
```suggestion
        // in this context, if user provides both page and item to showProp, page should win
```

**Review by halocline — COMMENTED** [review]
I like this a lot! Couple minor nits, then ready to merge.

**Review by halocline — APPROVED** [review]

**Pull request merged**